### PR TITLE
Added classes for async load balancer interceptor

### LINKED ISSUE #36: Support Ribbon in AsyncRestTemplate
See https://github.com/spring-cloud/spring-cloud-netflix/issues/233

see [MyfeedAutoConfig.java#L68-L83](https://github.com/spencergibb/myfeed/blob/master/myfeed-core/src/main/java/myfeed/MyfeedAutoConfig.java#L68-L83) and [RibbonAsyncClientHttpRequestFactory](https://github.com/spencergibb/myfeed/blob/master/myfeed-core/src/main/java/myfeed/RibbonAsyncClientHttpRequestFactory.java)

Not as straight forward as the standard synchronous `RestTemplate` because there is no concept of interceptors and URI resolution happens in various implementations of `AsyncClientHttpRequestFactory`

### PR DESCRIPTION
**Pull request by rworsnop**
Fixes #36

### CODE REVIEW

File: spring-cloud-commons/src/main/resources/META-INF/spring.factories [hunk 1]
```diff
@@ -4,6 +4,7 @@ org.springframework.cloud.client.CommonsClientAutoConfiguration,\
 org.springframework.cloud.client.discovery.noop.NoopDiscoveryClientAutoConfiguration,\
 org.springframework.cloud.client.hypermedia.CloudHypermediaAutoConfiguration,\
 org.springframework.cloud.client.loadbalancer.LoadBalancerAutoConfiguration,\
+org.springframework.cloud.client.loadbalancer.AsyncLoadBalancerAutoConfiguration,\
```
Comment (spencergibb) [inline spring-cloud-commons/src/main/resources/META-INF/spring.factories]: Alphabetical order, please.

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/AsyncLoadBalancerAutoConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+ *
+ * @author Rob Worsnop
+ */
+@Configuration
+@ConditionalOnClass(AsyncRestTemplate.class)
+@ConditionalOnBean(LoadBalancerClient.class)
+@EnableConfigurationProperties(LoadBalancerRetryProperties.class)
+public class AsyncLoadBalancerAutoConfiguration {
+
+    @LoadBalanced
```
Comment (spencergibb) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/AsyncLoadBalancerAutoConfiguration.java]: Formatting, we use tabs, not spaces

File: spring-cloud-commons/src/test/java/org/springframework/cloud/client/loadbalancer/AsyncLoadBalancerAutoConfigurationTests.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+import org.springframework.web.client.AsyncRestTemplate;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+
+import static org.hamcrest.Matchers.*;
```
Comment (spencergibb) [inline spring-cloud-commons/src/test/java/org/springframework/cloud/client/loadbalancer/AsyncLoadBalancerAutoConfigurationTests.java]: No `*` imports

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/AsyncLoadBalancerInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+	private LoadBalancerClient loadBalancer;
+
+	public AsyncLoadBalancerInterceptor(LoadBalancerClient loadBalancer) {
+		this.loadBalancer = loadBalancer;
+	}
+
+	@Override
+	public ListenableFuture<ClientHttpResponse> intercept(final HttpRequest request, final byte[] body,
+														  final AsyncClientHttpRequestExecution execution) throws IOException {
```
Comment (ryanjbaxter) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/AsyncLoadBalancerInterceptor.java]: weird line return here, or is it GitHub?

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/AsyncLoadBalancerInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+		final URI originalUri = request.getURI();
+		String serviceName = originalUri.getHost();
+		return this.loadBalancer.execute(serviceName,
```
Comment (ryanjbaxter) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/AsyncLoadBalancerInterceptor.java]: We recently added retry logic to the `RestTemplate` interceptor, should we do the same for the async interceptor?

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/AsyncLoadBalancerInterceptor.java (hunk 4 above)
Comment (rworsnop) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/AsyncLoadBalancerInterceptor.java]: Looks like browser tabs are worth more than IDE tabs. I've reduced the number of them.

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/AsyncLoadBalancerInterceptor.java (hunk 5 above)
Comment (rworsnop) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/AsyncLoadBalancerInterceptor.java]: I noticed that, but `RetryTemplate` does not seem to lend itself to asynchronous operations. Or am I missing something?

Comment (ryanjbaxter) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/AsyncLoadBalancerInterceptor.java]: That is a good point, clearly I didn't put to much thought into that comment when I made it :)  Let me ask around and see if anyone has done something like this yet

**Comment by rworsnop** [comment]
@spencergibb Will this be in Dalston.M1?

**Comment by spencergibb** [comment]
It's in my queue to look at. I think it will.

**Pull request merged**